File: Python/HashTable/BJAlgo_4195.py
from sys import stdin


def find(p):
    if p == parent[p]:
        return p
    else:
        pp = find(parent[p])
        parent[p] = pp
        return parent[p]


def union(p1, p2):
    parent1 = find(p1)
    parent2 = find(p2)

    if parent1 != parent2:
        parent[parent2] = parent1
        number[parent1] += number[parent2]


test_case = int(stdin.readline())
for _ in range(test_case):
    parent = dict()
    number = dict()
    F = int(stdin.readline())
    for _ in range(F):
        person1, person2 = stdin.readline().split()

        if person1 not in parent:
            parent[person1] = person1
            number[person1] = 1

        if person2 not in parent:
            parent[person2] = person2
            number[person2] = 1

        union(person1, person2)

        print(number[find(person1)])

File: Java/HashTable/BJAlgo_4195.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;

public class BJAlgo_4195 {
    static Map<String, String> parent = new HashMap<>();
    static Map<String, Integer> networks = new HashMap<>();

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int testCase = Integer.parseInt(br.readLine());
        for (int i = 0; i < testCase; i++) {
            parent.clear();
            networks.clear();

            int f = Integer.parseInt(br.readLine());
            for (int j = 0; j < f; j++) {
                String[] people = br.readLine().split(" ");

                if (!parent.containsKey(people[0])) {
                    parent.put(people[0], people[0]);
                    networks.put(people[0], 1);
                }

                if (!parent.containsKey(people[1])) {
                    parent.put(people[1], people[1]);
                    networks.put(people[1], 1);
                }

                union(people[0], people[1]);
                System.out.println(networks.get(find(people[0])));
            }
        }
    }

    private static String find(String p) {
        if (p.equals(parent.get(p))) {
            return p;
        }

        String par = find(parent.get(p));
        parent.put(p, par);
        return parent.get(p);
    }

    private static void union(String a, String b) {
        String parent1 = find(a);
        String parent2 = find(b);

        if (!parent1.equals(parent2)) {
            parent.put(parent2, parent1);
            networks.put(parent1, networks.get(parent1) + networks.get(parent2));
        }
    }
}
